### PR TITLE
ci(i): Use composites & split test matrix into smaller jobs

### LINKED ISSUE #3025: GH branch protection rules for required check is buggy with our long ci names.
GH branch protection rules for required check is buggy with our long ci names.

Currently whenever a new type of testing is introduced we append a new dimension to our existing matrix which causes the following:
- Someone has to remove all the existing required github checks and then re-add them as all the matrix names changes
- With our latest addition the github ui trips up further and uses `...` because of character limit this means it's impossible to select the exact job we want to make required in the protection rules.

If we split up the jobs logically into smaller jobs and matrix and make use of the github composite feature, we only need to add the new introduced jobs not have to remove and readd all previous ones, and can have more sensible coverage artifact names and job names to easily select in branch protection UI. 

For example here we can't see materialized option at all only `...`

image

This is worsened by the fact that github `...` also combined all clashing names into a globing like behavior (so selects more than one if the name is too long, you can't specify the exact job)

### PR DESCRIPTION
**Pull request by shahzadlone**
## Relevant issue(s)
Resolves #3025

### Problem
More about the problem in the linked issue #3025

### Changes
- Reusable composite actions are introduced under `.github/composites` folder.
- Still the same number of jobs as before.
- `badger-memory` name is now shown as `memory`
- `badger-file` name is now shown as `file`
- Split the basic 18 tests into a separate matric job.
- All `includes` tests we had before are now split in their own jobs.
- If a job fails, have more granular control now to re-run the job without waiting for all to finish.
- The job names won't clash with github ui with `...` anymore.
- The job names will remain static now.
- Upon new tests being introduced we don't have to change previous required branch protection rules, just add new ones.
- Upon new tests being introduced the author doesn't need to worry about making it required in the same PR.
- Easy to install sourcehub locally for any ci action now
- Easy to setup defradb with caching optimizations for any action now
- Change detector also now utilizes caching logic now, due to ease of use of composites.
- Updated the workflow name to be shorter as it looked odd in the readme before.

## How has this been tested?
- Example run: https://github.com/sourcenetwork/defradb/actions/runs/10914739872?pr=3027

### CODE REVIEW

**Comment by shahzadlone** [comment]
Example:

image

**Review by AndrewSisley — COMMENTED** [review]
Shahzad, this actually looks amazing, and a real surprise - where did you get the idea from?  Or had this been sitting in your head for a while?

I have a question regarding an action, but otherwise this looks really really good.

**Comment by shahzadlone** [comment]
> Shahzad, this actually looks amazing, and a real surprise - where did you get the idea from? Or had this been sitting in your head for a while?

I didn't like the extra list of `includes` list we had before in addition to the matrix. I knew we could break them into more sensible groups. Earlier in the week when we updated the branch protection rules for your view feature and I found the GitHub UI buggy I thought I can give it a try haha then the composite actions made it much cleaner than I expected.

File: .github/composites/test-coverage-with-artifact/action.yml
Comment (shahzadlone) [reply .github/composites/test-coverage-with-artifact/action.yml:33]: More like the matrix is running this composite action (bunch of steps like before, now shared instead of duplicating them in every matric job). The composite inherits the env variables from the parent action (so whatever env variables are set when this is called will be applied to them, you can also manually see the env variables for every jobs step when this is called and all the env variables it inherits in the run logs.)

Comment (AndrewSisley) [reply .github/composites/test-coverage-with-artifact/action.yml:33]: Ah got it, that must be the reference `uses: ./.github/composites/test-coverage-with-artifact` in the new jobs then I assume.

Thanks Shahzad :)

Comment (shahzadlone) [reply .github/composites/test-coverage-with-artifact/action.yml:33]: Yes, correct.

**Review by AndrewSisley — APPROVED** [review]
LGTM, thank you very much for this - I suggest writing a quick heads-up (no need to wait for their approval IMO) in the dev-db channel to make sure people are aware as it affects everyone.

**Pull request merged**
